server/routerlicious/packages/services-core/src/mongoCheckpointRepository.ts

```diff
@@ -0,0 +1,37 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ICollection, ICheckpointRepository } from "./database";
+import { ICheckpoint, IDeliState, IScribe } from "./document";
+
+export class MongoCheckpointRepository implements ICheckpointRepository {
+	constructor(private readonly collection: ICollection<ICheckpoint>, private readonly checkpointType: string) {
+    }
+
+    async getCheckpoint(documentId: string, tenantId: string): Promise<ICheckpoint> {
+        const pointReadFilter = this.composePointReadFilter({documentId, tenantId});
+        return this.collection.findOne(pointReadFilter);
+    }
+
+    async writeCheckpoint(documentId: string, tenantId: string, checkpoint: IDeliState|IScribe): Promise<void> {
+        const pointReadFilter = this.composePointReadFilter({documentId, tenantId});
+        await this.collection.upsert(pointReadFilter, {[this.checkpointType]: JSON.stringify(checkpoint)}, null)
+    }
+
+    async removeServiceCheckpoint(documentId, tenantId): Promise<void> {
+        const pointReadFilter = this.composePointReadFilter({documentId, tenantId});
+        await this.collection.upsert(pointReadFilter, {[this.checkpointType]: "" }, null);
+    }
+
+    async deleteCheckpoint(documentId: string, tenantId: string): Promise<void> {
+        const pointReadFilter = this.composePointReadFilter({documentId, tenantId});
+        await this.collection.deleteOne(pointReadFilter);
+    }
+
+	private composePointReadFilter(filter: any): { _id: string; documentId: string } & any {
+		const documentId = filter.documentId;
+		return { ...filter, _id: documentId };
+	}
```

Comment:
Alignment here is wrong.

Also, all your filter is `{documentId, tenantId}`, no need to create an extra filter. Lastly, learned something that `documentId+_id+tenantId` is not a point read filter....

I think you can do this:
```
private composePointReadFilter(documentId: string, tenantId:string): { _id: string; documentId: string }  {
		return {_id: documentId, documentId};
	}
```


Or change your model `_id` to be `documentId+tenantId`, then your composePointReadFilter would become (This might be better if no length restrications):
```
private composePointReadFilter(documentId: string, tenantId:string): { _id: string; documentId: string }  {
		return {_id: tenantId+documentId, documentId};
	}
```